application/common/models/User.php

```diff
@@ -521,22 +541,81 @@ public function getDisplayName()
         return $this->display_name ?? "$this->first_name $this->last_name";
     }
 
+    /**
+     * Based on provided time, determine whether to present a reminder to add
+     * an MFA option.
+     * @param int $now
+     * @return bool
+     * @throws Exception
+     */
+    public function isTimeToNagToAddMfa(int $now): bool
+    {
+        return MySqlDateTime::isBefore($this->nag_for_mfa_after, $now)
+            && (count($this->getVerifiedMfaOptions()) === 0);
+    }
+
+    /**
+     * Based on provided time, determine whether to present a reminder to add
+     * a recovery method option.
+     * @param int $now
+     * @return bool
+     * @throws Exception
+     */
+    public function isTimeToNagToAddMethod(int $now): bool
+    {
+        return MySqlDateTime::isBefore($this->nag_for_method_after, $now)
+            && (count($this->getVerifiedMethodOptions()) === 0);
+    }
+
     /**
      * Based on current time, determine whether to present a profile review to
      * the user.
+     * @param int $now
+     * @return bool
+     * @throws Exception
      */
-    public function isTimeForReview()
+    public function isTimeForReview(int $now)
     {
-        return MySqlDateTime::isBefore($this->review_profile_after, time());
+        return MySqlDateTime::isBefore($this->review_profile_after, $now);
     }
 
+    /**
+     * Based on current time and presence of MFA and Method options,
+     * determine which "nag" to present to the user.
+     *
+     */
+    public function getNagState()
+    {
+        /*
+         * Don't recalculate in case the date has changed since the last calculation.
+         */
+        if ($this->nagState !== null) {
+            return $this->nagState;
+        }
+        $possibleNags = [
+            self::NAG_ADD_MFA => 'isTimeToNagToAddMfa',
+            self::NAG_ADD_METHOD => 'isTimeToNagToAddMethod',
+            self::NAG_PROFILE_REVIEW => 'isTimeForReview',
+        ];
+        $now = time();
+        foreach ($possibleNags as $nagType => $isTime) {
+            if ($this->$isTime($now)) {
+                $this->nagState = $nagType;
+                return $this->nagState;
+            }
+        }
+        return self::NAG_NONE;
+    }
+
+
     /**
      * @return array MFA related properties
      */
     public function getMfaFields()
     {
         return [
             'prompt'  => $this->isPromptForMfa() ? 'yes' : 'no',
```

Comment:
are `prompt` and `add` mutually exclusive?

Reply:
No. That's what I though, too, at first. `prompt` tells `simplesamlphp-module-mfa` to present the user with a prompt for their *existing* mfa, whereas `add` tells `simplesamlphp-module-profilereview` to present the user with a reminder to add an `mfa` because they don't have one. Though they seem opposite, there are plenty of occasions when neither will be `yes`, simply because the `nag_for_mfa_after` date hasn't passed.

Reply:
👍